server/modules/routing/readconnroute/readconnroute.cc

```diff
@@ -367,11 +367,20 @@ static MXS_ROUTER_SESSION* newSession(MXS_ROUTER* instance, MXS_SESSION* session
                     candidate = ref;
                 }
             }
-            else if ((ref->connections + 1) / ref->server_weight
-                     < (candidate->connections + 1) / candidate->server_weight)
+            else
             {
-                /* ref has a better score. */
-                candidate = ref;
+                uint64_t score1 = (ref->connections + 1) / ref->server_weight;
+                uint64_t score2 = (candidate->connections + 1) / candidate->server_weight;
+                if (score1 < score2)
+                {
+                    /* ref has a better score. */
+                    candidate = ref;
+                }
+                else if (score1 == score2 && ref->lru_clock < candidate->lru_clock)
```

Comment:
How often do you see this happening? If it is rare, a random probability of picking one of the servers might be simpler and would remove the need for the logical clock. It wouldn't be a completely balanced probability distribution but it should at least resolve the "stalemate problem" that happens when the score is the same.

Reply:
Last week we saw an online instance (with 3k qps),  almost 87% queries go to same node, so it is not rare;

Reply:
This logic we had added to 2.2 in May，and it perform well, before that we almost receive ticket about this problem every week;

Reply:
If 87% of the queries go to only one node that would suggest that a non-default load balancing algorithm was in use. The default (LEAST_CURRENT_OPERATIONS) should spread out queries evenly across all connected servers.

In 2.4 we changed the code to always use the number of connections when creating new connections and the load balancing algorithm only for queries. Have you seen this with that version?

Reply:
No, most of our instances are 2.2, few are 2.3, no 2.4